code/API_definitions/kyc-tenure.yaml

```diff
@@ -0,0 +1,485 @@
+              value:
+                status: 400
+                code: INVALID_ARGUMENT
+                message: Client specified an invalid argument, request body or query param.
+            GENERIC_400_OUT_OF_RANGE:
+              description: Out of Range. Specific Syntax Exception used when a given field has a pre-defined range or a invalid filter criteria combination is requested
+              value:
+                status: 400
+                code: OUT_OF_RANGE
+                message: Client specified an invalid range.
+
+    Generic401:
+      description: Unauthorized
+      headers:
+        x-correlator:
+          $ref: "#/components/headers/X-Correlator"
+      content:
+        application/json:
+          schema:
+            $ref: "#/components/schemas/ErrorInfo"
+          examples:
+            GENERIC_401_UNAUTHENTICATED:
+              description: Request cannot be authenticated
+              value:
+                status: 401
+                code: UNAUTHENTICATED
+                message: Request not authenticated due to missing, invalid, or expired credentials.
+            GENERIC_401_AUTHENTICATION_REQUIRED:
+              description: New authentication is needed, authentication is no longer valid
+              value:
+                status: 401
+                code: AUTHENTICATION_REQUIRED
+                message: New authentication is required.
+
+    Generic403:
+      description: Forbidden
+      headers:
+        x-correlator:
+          $ref: "#/components/headers/X-Correlator"
+      content:
+        application/json:
+          schema:
+            $ref: "#/components/schemas/ErrorInfo"
+          examples:
+            GENERIC_403_PERMISSION_DENIED:
+              description: Permission denied. OAuth2 token access does not have the required scope or when the user fails operational security
+              value:
+                status: 403
+                code: PERMISSION_DENIED
+                message: Client does not have sufficient permissions to perform this action.
+            GENERIC_403_INVALID_TOKEN_CONTEXT:
+              description: Reflect some inconsistency between information in some field of the API and the related OAuth2 Token
+              value:
+                status: 403
+                code: INVALID_TOKEN_CONTEXT
+                message: "{{field}} is not consistent with access token."
+
+    Generic404:
+      description: Not found
+      headers:
+        x-correlator:
+          $ref: "#/components/headers/X-Correlator"
+      content:
+        application/json:
+          schema:
+            $ref: "#/components/schemas/ErrorInfo"
+          examples:
+            GENERIC_404_NOT_FOUND:
+              description: Resource is not found
+              value:
+                status: 404
+                code: NOT_FOUND
+                message: The specified resource is not found.
+            GENERIC_404_DEVICE_NOT_FOUND:
```

Comment:
I just took the standard errors defined by Commonalities and added then to the first YAML proposal. I agree that some find the term "DEVICE" confusing, but nonetheless we are using `phoneNumber` to identify what CAMARA calls the "device", but in our case means "device subscription". Just because we exclude IP addresses or the network access identifier does not change that.

So `DEVICE_NOT_FOUND` is the error code we should be returning when the "device identifier" (in our case, subscription identifier) provided by the API consumer is not one that the API provider knows anything about. `NOT_FOUND` is the generic code to return when the thing not found is not a "device".

We can maybe argue that the CAMARA concept of a "device" is too broad, and we should clearly separate "physical device" from "device subscription", and maybe have special rules for `phoneNumber` only. But that is a discussion to have in Commonalities.

Reply:
> I just took the standard errors defined by Commonalities and added then to the first YAML proposal. I agree that some find the term "DEVICE" confusing, but nonetheless we are using `phoneNumber` to identify what CAMARA calls the "device", but in our case means "device subscription". Just because we exclude IP addresses or the network access identifier does not change that.

I don't agree with this statement. The Commonalities documentation, both for the errors and the info.description section, refers to the device object, not the concept of device.

Here we are not using the phone number as a device identifier, we are using it as a subscription identifier. I think that the concept of "device subscription" is misleading in this context. Is there an example API in CAMARA where the device object is not referring to a physical device? I only know Device Status and Device Location. There is also Device Swap which doesn't use the device object.

Reply:
I was thinking the API applied to all networks, but remember now it was mobile networks only. So forget "device subscription", and let's define "mobile subscription" and "physical mobile device".

- So the `Device` object can be used to identify subject of the API, and this can be either the mobile subscription or the physical mobile device. Both meanings are possible.

- Similarly, the `phoneNumber` can be used to identify the subject of the API, and again can be either the mobile subscription or the physical mobile device. This is to be expected, as `phoneNumber` is also a field of the `Device` object.

- Whether "mobile subscription" or "physical mobile device" is intended as the subject is **entirely dependent on the context of the individual API**. For this API, it is "mobile subscription", but that could have been identified using a `Device` object. I thought that cumbersome for this use case, so proposed `phoneNumber` only.

- In general, **this API will be called using a 3-legged token**. So the API consumer will generally not call the API with either an explicit `phoneNumber` or (if it had been defined instead) and explicit `Device` object. Whether `phoneNumber` or `Device` object is specified in the YAML makes **no difference** to how the 3-legged token is obtained and used.
- We (as API definers) have **no idea** what specific identifier is associated with the access token by all possible API providers. It does not have to be MSISDN. It does not even have to be requested using MSISDN in the `login_hint`, as `ipport:` is entirely acceptable for generating an access token. Access tokens can be valid for for multiple APIs, and whether the end user identifier associated with the access token is used to identify the mobile subscription or the physical mobile device is entirely dependent on the context of each API that the access token is valid for.

- Because we must define the possibility to pass an explicit mobile subscription identifier as the subject in the case that a 2-legged access token is used (this will not be the normal case) we open ourselves up to 2 additional errors:
  - The API consumer uses a 3-legged token but then anyway specifies an explicit mobile subscription identifier
  - The API consumer uses a 2-legged token, but does not specify any explicit mobile subscription identifier
- These error cases were considered by Commonalities, who defined two new error codes:
  - `DEVICE_IDENTIFIERS_MISMATCH` when a 3-legged token is provided but also an explicit identifier for a different mobile subscription or physical mobile device
  - `UNIDENTIFIABLE_DEVICE` when a 2-legged token is provided, but no  explicit identifier for a different mobile subscription or physical mobile device

Now Commonalities have not documented this as well or as comprehensively as it could be, but the important point is that, whether the intended subject of the API is the mobile subscription or the physical mobile device, the **error is the same**, and hence **the same error code should be used**.

Defining new error codes just because the explicit subject identifier is restricted to `phoneNumber` rather than being the more general `Device` object, or because of an objection to using the term "DEVICE_" when the subject is a mobile subscription is pointless, and increases implementation complexity (for both API consumer and provider).

The error `message` can always be customised, as these are just examples, and so I am happy to define more relevant error messages. But new error codes not defined by Commonalities should go through Commonalities first, and not just be introduced by APIs arbitrarily.

In this case, I think we could remove all objections by asking Commonalities to rename the error codes to something like `DEVICE_OR_SUBSCRIPTION_IDENTIFIER_MISMATCH` and `UNIDENTIFIABLE_DEVICE_OR_SUBSCRIPTION` respectively.

Reply:
Hi Eric,

I agree the same error code should be used, but I think the use of the word device is leading to confusion. because people may think this refers to hardware.

One suggestion from my side : It may be a better idea to use for example IDENTIFIER_MISMATCH and MISSING_IDENTIFIER   in stead of DEVICE_IDENTIFIERS_MISMATCH and UNIDENTIFIABLE_DEVICE.  This keeps it in the middle what identifier it is, and also does not point to a device or sim card.

Reply:
I think we have a different vision on what has been define in Commonalities regarding the device. 
The discussion originated from the need to standarize the errors across all APIs that use the `device object` explicitly.
Your arguments cover a semantic vision that extends the behaviour. I'm not saying that is wrong, because I understand what you are trying to explain, but in my opinion it is not what the commonalities discussion intended and there is a gap in the understanding of the device that maybe we are not transferring good enough in the documentation. And this gap might affect also developers that may see the concept device and don't know what are we exactly referring to.

> I think this is something the Commonalities will resolve in due course. We can't have a situation where APIs arbitrarily define new error codes just because they don't like the name of the existing ones.

I agree, this should be discussed in Commonalities, but the WGs are not defining arbitrarily, there has been a process to reach that conclusion and it complies with the API design guidelines.

Reply:
The Commonalities guidelines give no guidance on using `404 NOT_FOUND`. But I agree we do not need both `NOT_FOUND` and `DEVICE_NOT_FOUND`, so one has to go. I don't see that we need to decide it for this initial WIP version, however.

Discussing whether a `phoneNumber` is a device identifier is more than semantics, because deciding that it is not means that different APIs need to return different errors for what is the same logical error. This complicates implementation, and I think will also ultimately confuse application developers who need to develop this parallel but otherwise common logic to handle "device" and "phone number" errors.

Reply:
See [this](https://github.com/camaraproject/Commonalities/issues/321#issuecomment-2429158445) comment. Proposal would be `IDENTIFIER_NOT_FOUND`, with `message` giving details.

Reply:
hi @eric-murray , @fernandopradocabrillo , as discussed with Fernando last week we must also add normalized error status HTTP-422 in kyc-match which are not present. It must consistent with what will be decided for Tenure API.

Thanks
BR
Gilles

Reply:
@GillesInnov35
Ultimately, this will be decided by Commonalities PR [#316](https://github.com/camaraproject/Commonalities/pull/316). KYC cannot decide such error code independently of Commonalities.

These are the 422 codes proposed by that PR:
```
                  code:
                    enum:
                      - IDENTIFIER_MISMATCH
                      - SERVICE_NOT_APPLICABLE
                      - UNIDENTIFIABLE_IDENTIFIER
                      - UNSUPPORTED_IDENTIFIER
                      - UNNECESSARY_IDENTIFIER
                      - "{{SPECIFIC_CODE}}"
 ```

{{SPECIFIC_CODE}} is for any scenario not considered by Commonalities, but does not give API sub-projects free rein to ignore the defined 422 codes.

Reply:
thanks Eric. Currently at Orange we need those 2 bellow errors status avalaible in Sim Swap API in order to have consistent development. Hoping decision will be soon.

>       GENERIC_422_NOT_SUPPORTED:
              description: Not Supported
              value:
                status: 422
                code: NOT_SUPPORTED
                message: Service not supported for this phoneNumber
            UNIDENTIFIABLE_PHONE_NUMBER:
              description: The phone number is not included in the request and the phone number information cannot be derived from the 3-legged access token
              value:
                status: 422
                code: UNIDENTIFIABLE_PHONE_NUMBER
                message: The phone number cannot be identified

BR
Gilles

Reply:
sorry Eric, my comment was not clear. Those 2 errors with associated messages are already available in Simswap and to have consistent API development we need also 422 error.

Reply:
> Those 2 errors with associated messages are already available in Simswap and to have consistent API development we need also 422 error.

SIM Swap should have complied with CAMARA Commonalities guidelines on [Error Responses](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md#6-error-responses), and that requirement will remain as the guidelines are updated.

So SIM Swap will need to update its error response `code`s (but not necessarily the error `message`) in any future version to be CAMARA compliant. This is a consequence of SIM Swap defining custom error messages and not engaging with Commonalities. But there is no obligation to update - SIM Swap can continue as v1.0.0 forever.

But we can't have the situation where new APIs unrelated to SIM Swap maintain an ongoing non-compliance with CAMARA guidelines solely to simplify the development tracks of some API providers.

Reply:
Hi @eric-murray @GillesInnov35 
I also feel more comfortable aligning with what was decided for Sim Swap. I don't share the opinion that it was not compliant with commonalities, it was not a random decision, there was a reasoning behind. 
I like the path that this issue is taking in commonalities but maybe waiting for it to be formilize in Spring25 is too long.

Reply:
hi @eric-murray @fernandopradocabrillo, I think we should made a decision before Sring25.
Thanks

BR
Gilles

Reply:
@fernandopradocabrillo
OK. I'll add the 422 codes `NOT_SUPPORTED` and `UNIDENTIFIABLE_PHONE_NUMBER`, and then we can review once the next release of the Commonalities guidelines is available.

Reply:
> I think we should made a decision before Sring25. Thanks

A release candidate for the next version of the Commonalities guidelines will be available soon (due 7/11, but that will be missed), and final version by 7/12.

Reply:
hi @fernandopradocabrillo , @eric-murray, I understand that we could include the new commonalities errors (see bellow). Good, it seems ok. I'll check quickly with Ludovic Robert which is involved in the technical meetings.

I guess the error code names would not be a problem for the product team, actually what is most important is the error status.

>  422        |     UNSUPPORTED_IDENTIFIER | The identifier provided is not supported.
  422        |       IDENTIFIER_MISMATCH  | Provided identifiers are not consistent.
  422        |     SERVICE_NOT_APPLICABLE | The service is not available for the provided identifier. 
  422       |      MISSING_IDENTIFIER  | The device cannot be identified.

Fernando, do you think we could now also apply those commonalities into the kyc Match definition ?

Thanks a lot
Gilles

Reply:
For KYC I'm sure I prefer the commonalities outcome because we won't create any release until the Spring25 so it makes sense to align from the beginning.
I agree with @eric-murray, the errors seem to be almost definitive so worst case is to make small adjustments.

Reply:
hi @fernandopradocabrillo, so let's stay on this kyc-match version without any update of the errors section waiting for the Release Spring25 of Commonalities.
Thanks
Gilles

Reply:
@fernandopradocabrillo, @GillesInnov35, @grgpapadopoulos 
> just to finish with this topic and continue with the PR. Yesterday in the meeting we were talking about this alignment and what errors to include. Initially I was more prone to align with sim swap, but maybe, if it is stable enough, we can include the new commonalities errors so that way we don't have to do another version for Spring25. From TEF, the product team hopes that we could launch a release by the end of the month. WDYT?

I updated the PR to include the relevant 404 and 422 codes in Commonalities PR #[329](https://github.com/camaraproject/Commonalities/pull/329). That PR is not merged yet, but I do not expect significant changes. I anyway used custom error messages, so hopefully the reason why each code might be returned is clear.

I think we can do a "0.1.0-alpha.1" release by the end of the month. For a 0.1.0 release, we really need <i>some</i> test cases defined, and would also need to apply any updated Commonalities or ICM rules that are defined for this meta-release cycle. None of that should change the API definition significantly, so there is still value in an alpha release.